agent/hcp/telemetry_provider.go

```diff
@@ -0,0 +1,184 @@
+type hcpProviderImpl struct {
+	// cfg holds configuration that can be dynamically updated.
+	cfg *dynamicConfig
+	// cfgHash is used to check if two dynamicConfig objects are equal.
+	cfgHash uint64
+
+	// A reader-writer mutex is used as the provider is read heavy.
+	// OTEL components access telemetryConfig during metrics collection and export (read).
+	// Meanwhile, config is only updated when there are changes (write).
+	rw sync.RWMutex
+	// hcpClient is an authenticated client used to make HTTP requests to HCP.
+	hcpClient hcpclient.Client
+}
+
+// dynamicConfig is a set of configurable settings for metrics collection, processing and export.
+type dynamicConfig struct {
+	endpoint *url.URL
+	labels   map[string]string
+	filters  *regexp.Regexp
+	// refreshInterval controls the interval at which configuration is fetched from HCP to refresh config.
+	refreshInterval time.Duration
+}
+
+// providerParams is used to initialize a hcpProviderImpl.
+type providerParams struct {
+	metricsConfig   *hcpclient.MetricsConfig
+	refreshInterval time.Duration
+	hcpClient       hcpclient.Client
+}
+
+// NewHCPProviderImpl initializes and starts a HCP Telemetry provider with provided params.
+func NewHCPProviderImpl(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
+	if params.hcpClient == nil {
+		return nil, fmt.Errorf("missing HCP client")
+	}
+
+	if params.metricsConfig == nil {
+		return nil, fmt.Errorf("missing metrics config")
+	}
+
+	if params.refreshInterval <= 0 {
+		return nil, fmt.Errorf("invalid refresh interval")
+	}
+
+	cfg := &dynamicConfig{
+		endpoint:        params.metricsConfig.Endpoint,
+		labels:          params.metricsConfig.Labels,
+		filters:         params.metricsConfig.Filters,
+		refreshInterval: params.refreshInterval,
+	}
+
+	hash, err := calculateHash(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate hash: %w", err)
+	}
+
+	t := &hcpProviderImpl{
+		cfg:       cfg,
+		cfgHash:   hash,
+		hcpClient: params.hcpClient,
+	}
+
+	go t.run(ctx)
+
+	return t, nil
+}
+
+// run continously checks for updates to the telemetry configuration by making a request to HCP.
+// Modification of config only occurs if changes are detected to decrease write locks that block read locks.
+func (t *hcpProviderImpl) run(ctx context.Context) {
+	ticker := time.NewTicker(t.cfg.refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if newCfg, hasChanged := t.checkUpdate(ctx); hasChanged {
+				t.modifyTelemetryConfig(newCfg)
+				ticker.Reset(newCfg.refreshInterval)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// checkUpdate makes a HTTP request to HCP to return a new metrics configuration and true, if config changed.
+// checkUpdate does not update the metricsConfig field to prevent acquiring the write lock unnecessarily.
+func (t *hcpProviderImpl) checkUpdate(ctx context.Context) (*dynamicConfig, bool) {
+	t.rw.RLock()
+	defer t.rw.RUnlock()
```

Comment:
this lock is taken out to read the existing config? Can we just wrap the grabbing of `t.cfgHash`? This right now wraps the entire network call to HCP which will takes ages in lock terms

Reply:
yep thanks for catching! moved it down